### PR TITLE
Add StartWhenSynchronized option for RPC server

### LINKED ISSUE #2433: Option to start RPC after chain sync
We need an option to enable RPC service at the `neo-go` start only after the chain sync is done. 

NeoFS need this to prevent processing the irrelevant notifications from the past. Other dApps may find this option useful for similar reasons.

### PR DESCRIPTION
**Pull request by roman-khimov**
Not completely tested yet, fixes #2433 and adds some building blocks for #1949.

### CODE REVIEW

**Comment by roman-khimov** [comment]
Seems to work fine.

File: pkg/rpc/server/server.go [hunk 1]
```diff
@@ -66,6 +67,7 @@ type (
 		log              *zap.Logger
 		https            *http.Server
 		shutdown         chan struct{}
+		started          *atomic.Bool
```
Comment (fyrchik) [inline pkg/rpc/server/server.go:70]: Why is it a pointer?

File: cli/server/server.go [hunk 2]
```diff
@@ -528,20 +528,14 @@ Main:
...
 			case syscall.SIGHUP:
 				log.Info("SIGHUP received, restarting rpc-server")
-				serverErr := rpcServer.Shutdown()
-				if serverErr != nil {
-					errChan <- fmt.Errorf("error while restarting rpc-server: %w", serverErr)
-					break
-				}
-				rpcServer = server.New(chain, cfg.ApplicationConfiguration.RPC, serv, oracleSrv, log)
-				rpcServer.Start(errChan)
+				rpcServer.Shutdown()
```
Comment (fyrchik) [inline cli/server/server.go:534]: It seems there is a potential deadlock here, because we have an unbuffered channel which is read in the same `select`. Am I missing something?

File: cli/server/server.go [hunk 3]
```diff
@@ -530,12 +533,14 @@ Main:
...
-				rpcServer.Start()
+				serv.AddService(&rpcServer) // Replaces old one by service name.
+				if !cfg.ApplicationConfiguration.RPC.StartWhenSynchronized || serv.IsInSync() {
+					rpcServer.Start()
+				}
 			}
 		case <-grace.Done():
 			signal.Stop(sighupCh)
 			serv.Shutdown()
-			rpcServer.Shutdown()
```
Comment (fyrchik) [inline cli/server/server.go]: Why is this line removed? I thought `started` check in `shutdown` fixed any possible issues?

File: cli/server/server.go [hunk 4]
```diff
@@ -530,12 +533,14 @@ Main:
 				log.Info("SIGHUP received, restarting rpc-server")
 				rpcServer.Shutdown()
 				rpcServer = server.New(chain, cfg.ApplicationConfiguration.RPC, serv, oracleSrv, log, errChan)
-				rpcServer.Start()
+				serv.AddService(&rpcServer) // Replaces old one by service name.
+				if !cfg.ApplicationConfiguration.RPC.StartWhenSynchronized || serv.IsInSync() {
+					rpcServer.Start()
+				}
```
Comment (fyrchik) [inline cli/server/server.go:539]: So it doesn't start automatically, only on `SIGHUP`?

File: cli/server/server.go (hunk 3 above)
Comment (roman-khimov) [reply cli/server/server.go]: `network.(*Server)` stops all services.

File: pkg/rpc/server/server.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/rpc/server/server.go:70]: It doesn't matter much, but historically we return `Server` from `New` and not `*Server`.

File: cli/server/server.go (hunk 2 above)
Comment (roman-khimov) [reply cli/server/server.go:534]: That very well might be the case, although nothing changed wrt it in this PR.

File: pkg/rpc/server/server.go [hunk 5]
```diff
@@ -224,55 +225,56 @@ func (s *Server) Start(errChan chan error) {
...
+	s.log.Info("shutting down RPC server", zap.String("endpoint", s.Addr))
 	err := s.Server.Shutdown(context.Background())
 
 	// Wait for handleSubEvents to finish.
 	<-s.executionCh
 
-	if err == nil {
-		return httpsErr
+	if httpsErr != nil {
+		s.errChan <- httpsErr
```
Comment (AnnaShaleva) [inline pkg/rpc/server/server.go:281]: Shouldn't we return after that? Otherwice we need to be sure that the one who handles errors from `errChan` is able to receive two errors in a raw (both `httpsErr` and `err`) if so.

File: cli/server/server.go (hunk 4 above)
Comment (roman-khimov) [reply cli/server/server.go:539]: Either it's in the default configuration (StartWhenSynchronized false) and it works the way it worked before, or it's configured to StartWhenSynchronized and it only starts when the node is synchronized.

File: cli/server/server.go [hunk 6]
```diff
@@ -504,11 +504,11 @@ func startServer(ctx *cli.Context) error {
 	if err != nil {
 		return cli.NewExitError(err, 1)
 	}
-	rpcServer := server.New(chain, cfg.ApplicationConfiguration.RPC, serv, oracleSrv, log)
 	errChan := make(chan error)
+	rpcServer := server.New(chain, cfg.ApplicationConfiguration.RPC, serv, oracleSrv, log, errChan)
 
 	go serv.Start(errChan)
```
Comment (AnnaShaleva) [inline cli/server/server.go:511]: It's good that we provide `errChan` to RPC server constructor, but currently RPC server performs writes to `errChan` and doesn't perform reads. `serv` also doesn't read from `errChan`. I think, it should be fixed.

File: pkg/services/stateroot/validators.go [hunk 7]
```diff
@@ -17,6 +17,11 @@ const (
 	firstVoteResendDelay = 3 * time.Second
 )
 
+// Name returns service name.
+func (s *service) Name() string {
+	return "consensus"
```
Comment (AnnaShaleva) [inline pkg/services/stateroot/validators.go:22]: state validation?

File: docs/node-configuration.md [hunk 8]
```diff
@@ -158,6 +159,10 @@ where:
 - `MaxNEP11Tokens` - limit for the number of tokens returned from
   `getnep11balances` call.
 - `Port` is an RPC server port it should be bound to.
+- `StartWhenSynchronized` controls when RPC server will be started, by default
+  (`false` setting) it's started immediately and RPC is availabe during node
+  synchronization. Setting it to `true` will make the node start RPC service only
+  after full synchronization.
```
Comment (AnnaShaleva) [inline docs/node-configuration.md:165]: Probably, it should be mentioned here that RPC service will be started after the node reaches 2/3 of the connected peers height.

File: cli/server/server.go (hunk 6 above)
Comment (AnnaShaleva) [reply cli/server/server.go:511]: I've just noticed that we handle errors in a separate Main loop, so it's OK. But https://github.com/nspcc-dev/neo-go/pull/2445#discussion_r857334777 is still relevant.

File: cli/server/server.go (hunk 2 above)
Comment (roman-khimov) [reply cli/server/server.go:534]: Fixed.

File: pkg/rpc/server/server.go (hunk 5 above)
Comment (roman-khimov) [reply pkg/rpc/server/server.go]: Fixed.

File: pkg/services/stateroot/validators.go (hunk 7 above)
Comment (roman-khimov) [reply pkg/services/stateroot/validators.go]: Fixed.

File: docs/node-configuration.md (hunk 8 above)
Comment (roman-khimov) [reply docs/node-configuration.md:165]: It affects all of the services, so I've added another section.

**Review by fyrchik — APPROVED** [review]

**Review by AnnaShaleva — APPROVED** [review]
